File: clients/python/awan.py
"""Drive the `awan` character from Python — a thin wrapper over the binary.

The binary is the integration surface, so there is no SDK to speak of: every
function here just spawns `awan`. Install the binary first (any of):

    npm i -g awan          # prebuilt, no toolchain
    cargo install awan-cli
    # or download from https://github.com/codewithwan/awan/releases

See docs/INTEGRATE.md for the full event vocabulary.
"""

from __future__ import annotations

import shutil
import subprocess
from typing import Optional

BIN = shutil.which("awan") or "awan"


def _char(character: Optional[str]) -> list[str]:
    return ["-c", character] if character else []


def react(event: str, character: Optional[str] = None) -> None:
    """Play the one-shot reaction to *event* (e.g. "task.done"), then return."""
    subprocess.run([BIN, "react", event, *_char(character)], check=False)


def busy(label: str, character: Optional[str] = None) -> subprocess.Popen:
    """Start the "working…" loop with *label*; call .terminate() when done."""
    return subprocess.Popen([BIN, "busy", label, *_char(character)])


class Watch:
    """An ambient companion you feed events to over time.

    with Watch() as buddy:
        buddy.emit("cmd.start")
        buddy.emit("cmd.ok")
    """

    def __init__(self, character: Optional[str] = None) -> None:
        self._p = subprocess.Popen([BIN, "watch", *_char(character)], stdin=subprocess.PIPE)

    def emit(self, event: str) -> None:
        assert self._p.stdin is not None
        self._p.stdin.write(f"{event}\n".encode())
        self._p.stdin.flush()

    def stop(self) -> None:
        self._p.terminate()

    def __enter__(self) -> "Watch":
        return self

    def __exit__(self, *_exc) -> None:
        self.stop()

File: clients/go/awan.go
// Package awan drives the `awan` terminal character from Go. It shells out to
// the binary — the same personality layer every language reaches over the
// process boundary — so there is nothing to link. Install the binary with
// `npm i -g awan`, `cargo install awan-cli`, or from the GitHub Releases page.
//
// See docs/INTEGRATE.md for the full event vocabulary.
package awan

import "os/exec"

// bin is resolved from PATH; override it if you ship the binary elsewhere.
var bin = "awan"

func charArgs(character string) []string {
	if character != "" {
		return []string{"-c", character}
	}
	return nil
}

// React plays the one-shot reaction to event (e.g. "task.done"), then returns.
func React(event, character string) error {
	args := append([]string{"react", event}, charArgs(character)...)
	c := exec.Command(bin, args...)
	c.Stdout, c.Stderr = nil, nil
	return c.Run()
}

// Busy starts the "working…" loop with label. Call Cmd.Process.Kill when done.
func Busy(label, character string) (*exec.Cmd, error) {
	args := append([]string{"busy", label}, charArgs(character)...)
	c := exec.Command(bin, args...)
	return c, c.Start()
}

// Watch is an ambient companion you feed events to over time.
type Watch struct {
	cmd *exec.Cmd
	in  interface{ Write([]byte) (int, error) }
}

// StartWatch launches the companion. Emit events with (*Watch).Emit, end it
// with (*Watch).Stop.
func StartWatch(character string) (*Watch, error) {
	args := append([]string{"watch"}, charArgs(character)...)
	c := exec.Command(bin, args...)
	pipe, err := c.StdinPipe()
	if err != nil {
		return nil, err
	}
	if err := c.Start(); err != nil {
		return nil, err
	}
	return &Watch{cmd: c, in: pipe}, nil
}

// Emit pushes one event to the running companion.
func (w *Watch) Emit(event string) error {
	_, err := w.in.Write([]byte(event + "\n"))
	return err
}

// Stop ends the companion.
func (w *Watch) Stop() error { return w.cmd.Process.Kill() }
